Send Solana Tokens

memoriesadrift opened the pull request:
This PR is a mess due to being based off of 2 branches.  

**Start reviewing at `fefcc55a3318f429a3c795cf467104973ae5d1b1`**

## Description
Users can send tokens to arbitrary accounts, creating associated token accounts for receivers if need be.

What doesn't work:
Sending tokens to a token account holding a different token isn't currently handled. -- instead, it fails gracefully with the error message "Transaction signing error: failed to send transaction: Transaction simulation failed: Error processing Instruction 0: custom program error: 0x3"

memoriesadrift committed:
feat(suite): send raw Solana transactions

memoriesadrift committed:
chore(suite): add `@solana/web3.js` dependency

memoriesadrift committed:
chore(suite-common): add solana fee unit, chain ID, solana to lamports utility function

memoriesadrift committed:
chore(wallet-utils): bump `trezor-address-validator` to 0.4.4

memoriesadrift committed:
fix(blockchain-link): get latest blockhash instead of blockheight

memoriesadrift committed:
feat(suite): sign Solana transactions

memoriesadrift committed:
feat(connect): add default Solana fee info

memoriesadrift committed:
feat(blockchain-link): add `estimateFee` for Solana

memoriesadrift committed:
feat(suite): compose Solana transactions

memoriesadrift committed:
fix(components): display correct fee related values for Solana

memoriesadrift committed:
fix(suite): fix Solana send form test

memoriesadrift committed:
fix(suite): remove unused chainId for Solana

memoriesadrift committed:
feat(suite): dynamically load `@solana/web3.js` library

memoriesadrift committed:
feat(suite): display Solana tokens

memoriesadrift committed:
feat(blockchain-link): display whitelisted token names and symbols

memoriesadrift committed:
feat(blockchain-link): fetch Solana token whitelist

memoriesadrift committed:
squash! feat(blockchain-link): display whitelisted token names and symbols

memoriesadrift committed:
fix(blockchain-link): explicitly set WSOL symbol

memoriesadrift committed:
fix(blockchain-link): fix prettier issues

memoriesadrift committed:
fix(blockchain-link-utils): account for multiple token accounts holding
the same token

memoriesadrift committed:
squash! fix(suite): remove unused chainId for Solana

PeterBenc commented on packages/suite/src/actions/wallet/send/sendFormSolanaActions.ts:
```diff
@@ -75,38 +85,93 @@ const calculate = (
+                  account.symbol,
+                  tokenInfo.contract,
+              )
+            : [undefined, undefined];
```
This `[undefined, undefined]` seems unnecessary and not very nice,  so I would remove `fetchAccountOwnerAndTokenInfoForAddress` and just do 

```
    const accountInfoResponse = await TrezorConnect.getAccountInfo({
        coin: symbol,
        descriptor: address,
    });

    const accountOwner = accountInfoResponse.payload?.misc?.owner;
    const tokenInfo = accountInfoResponse.payload?.tokens?.find(
            token => token.contract === mint,
        )?.accounts;
```

Also btw isn't accountInfo already in `state` ? Do we need to refetch it here?

memoriesadrift replied:
this is for the *recipient*, this account comes from the form field. The `[undefined, undefined]` part is a fallback so that we don't fetch when we're not sending tokens == don't need this data.

PeterBenc commented on packages/suite/package.json:
```diff
@@ -54,6 +54,7 @@
         "@trezor/analytics": "workspace:*",
         "@trezor/blockchain-link": "workspace:*",
         "@trezor/blockchain-link-types": "workspace:*",
+        "@trezor/blockchain-link-utils": "workspace:*",
```
Is there a way around doing this? I can imagine trezor guys not being happy about this

memoriesadrift replied:
It depends on how you want to organise the code. It's either duplication or this. I don't know what trezor guys will prefer, so I just did it one way and want to wait and see what they say.

PeterBenc commented on packages/blockchain-link-types/src/params.ts:
```diff
@@ -29,8 +29,9 @@ export interface EstimateFeeParams {
-        data?: string; // eth tx data
+        data?: string; // eth tx data, sol tx message
         value?: string; // eth tx amount
+        createsAccount?: boolean; // sol account creation
```
I would try to be more specific here and also use `is` as prefix which is nice for flags like this because you do not need to parse the booliness of it from the whole name. e.g. `isTxCreatingTokenAccount`

PeterBenc commented on packages/connect/src/data/defaultFeeLevels.ts:
```diff
@@ -113,7 +113,7 @@ const SOLANA_FEE_INFO: FeeInfoWithLevels = {
     defaultFees: [{ label: 'normal', feePerUnit: '5000', blocks: -1 }],
     minFee: 5000,
-    maxFee: 5000,
+    maxFee: 100000000, // arbitrary
```
Is this a required parameter? if not I would drop it so it does not kick our ass in the future. What is the purpose of this field anyway?

memoriesadrift replied:
it is mandatory and since it's uncapped I plan to bring this up to the trezor guys in review

PeterBenc reviewed (commented):
Pretty much LGTM, pls check the comments

memoriesadrift committed:
feat(suite): create Solana token transfer instructions

memoriesadrift committed:
feat(suite): get required token accounts for solana token transfers

memoriesadrift committed:
feat(suite): create solana token account instructions

memoriesadrift committed:
feat(suite): build solana token transfer tx

memoriesadrift committed:
feat(blockchain-link): add token accounts to `TokenInfo`

memoriesadrift committed:
feat(suite): compose and sign solana token txs

memoriesadrift committed:
feat(suite): calculate solana token fees correctly

By taking into account the fee to fund creating recipient token accounts when necessary

memoriesadrift committed:
feat(suite): add solana token transfer tests
